src/shims/unix/unnamed_socket.rs

```diff
@@ -114,14 +114,32 @@ impl FileDescription for AnonSocket {
                     // Since there is no ErrorKind for EAGAIN, WouldBlock is used.
                     return ecx.set_last_error_and_return(ErrorKind::WouldBlock, dest);
                 } else {
+                    // TODO: move this to helper
                     // Blocking socketpair with writer and empty buffer.
-                    // FIXME: blocking is currently not supported
-                    throw_unsup_format!("socketpair/pipe/pipe2 read: blocking isn't supported yet");
+                    let peer_fd = self.peer_fd().clone();
+                    let dest = dest.clone();
+                    let weak_self_ref = self_ref.downgrade();
+                    ecx.block_thread(
+                        BlockReason::UnnamedSocket,
+                        None,
+                        callback!(
+                            @capture<'tcx> {
+                                weak_self_ref: WeakFileDescriptionRef,
+                                peer_fd: WeakFileDescriptionRef,
+                                len: usize,
+                                ptr: Pointer,
+                                dest: MPlaceTy<'tcx>,
+                            }
+                            @unblock = |this| {
+                                // TODO: We might need to decide what to do if peer_fd is closed when read is blocked.
+                                anonsocket_read(weak_self_ref, peer_fd, len, ptr, &dest, this)
+                            }
+                        ),
+                    );
                 }
             }
         }
-        // TODO: We might need to decide what to do if peer_fd is closed when read is blocked.
-        anonsocket_read(self, self.peer_fd().upgrade(), len, ptr, dest, ecx)
```

Comment:
You need to keep doing this here for in the case where we don't need to block

Reply:
Yea this is caught by the testcases too, it is now fixed :D